ethexe/network/src/db_sync/mod.rs

```diff
@@ -334,42 +159,149 @@ impl Config {
     }
 }
 
-type InnerBehaviour = request_response::Behaviour<ParityScaleCodec<Request, Response>>;
+#[async_trait]
+pub trait ExternalDataProvider: Send + Sync {
+    fn clone_boxed(&self) -> Box<dyn ExternalDataProvider>;
+
+    async fn programs_code_ids_at(
+        self: Box<Self>,
+        program_ids: BTreeSet<ActorId>,
+        block: H256,
+    ) -> anyhow::Result<Vec<CodeId>>;
+
+    async fn codes_states_at(
+        self: Box<Self>,
+        code_ids: BTreeSet<CodeId>,
+        block: H256,
+    ) -> anyhow::Result<Vec<CodeState>>;
+}
+
+#[derive(Debug, Eq, PartialEq, Copy, Clone, Hash)]
+pub struct RequestId(pub(crate) u64);
+
+#[derive(Debug, Eq, PartialEq, Copy, Clone, Hash)]
+pub struct ResponseId(pub(crate) u64);
+
+#[derive(derive_more::Debug, Default, Clone, Eq, PartialEq, Encode, Decode, derive_more::From)]
+pub struct HashesRequest(
+    #[debug("{:?}", AlternateCollectionFmt::set(_0, "hashes"))] pub BTreeSet<H256>,
+);
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ProgramIdsRequest {
+    pub at: H256,
+    pub expected_count: u64,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ValidCodesRequest {
+    pub at: H256,
+    pub validated_count: u64,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, derive_more::From)]
+pub enum Request {
+    Hashes(HashesRequest),
+    ProgramIds(ProgramIdsRequest),
+    ValidCodes(ValidCodesRequest),
+}
+
+impl Request {
+    pub fn hashes(request: impl Into<BTreeSet<H256>>) -> Self {
+        Self::Hashes(HashesRequest(request.into()))
+    }
+
+    pub fn program_ids(at: H256, expected_count: u64) -> Self {
+        Self::ProgramIds(ProgramIdsRequest { at, expected_count })
+    }
+
+    pub fn valid_codes(at: H256, validated_count: u64) -> Self {
+        Self::ValidCodes(ValidCodesRequest {
+            at,
+            validated_count,
+        })
+    }
+}
+
+#[derive(derive_more::Debug, Clone, Eq, PartialEq, derive_more::From, derive_more::Unwrap)]
+pub enum Response {
+    Hashes(#[debug("{:?}", AlternateCollectionFmt::map(_0, "entries"))] BTreeMap<H256, Vec<u8>>),
+    ProgramIds(
+        #[debug("{:?}", AlternateCollectionFmt::map(_0, "programs"))] BTreeMap<ActorId, CodeId>,
+    ),
+    ValidCodes(#[debug("{:?}", AlternateCollectionFmt::set(_0, "codes"))] BTreeSet<CodeId>),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct InnerProgramIdsRequest {
+    at: H256,
+}
+
+/// Network-only type to be encoded-decoded and sent over the network
```

Comment:
pub(crate) than?

Reply:
`InnerRespone` -> `InnerBehaviour` -> `pub Behaviour` -> `<Behaviour as NetworkBehaviour>::ConnectionHandler == THandler<InnerBehaviour>` and thus we have to keep it public